File: src/MagnusAddon/note/tags.py
from __future__ import annotations

from autoslot import Slots
from note.tag import Tag

f_root = "-::"
f_sentence: str = f"{f_root}sentence::"
f_kanji: str = f"{f_root}kanji::"
f_sentence_uses: str = f"{f_sentence}uses::"
f_vocab: str = f"{f_root}vocab::"
f_vocab_matching: str = f"{f_vocab}matching::"
f_vocab_register: str = f"{f_vocab}register::"
f_vocab_matching_requires: str = f"{f_vocab_matching}requires::"
f_vocab_matching_forbids: str = f"{f_vocab_matching}forbids::"
f_vocab_matching_todo: str = f"{f_vocab_matching}todo::"
f_vocab_matching_uses: str = f"{f_vocab_matching}uses::"
f_source = "source::"

class Tags(Slots):
    class Sentence(Slots):
        class Uses(Slots):
            incorrect_matches: Tag = Tag.from_name(f"{f_sentence_uses}incorrect-matches")
            hidden_matches: Tag = Tag.from_name(f"{f_sentence_uses}hidden-matches")

    class Kanji(Slots):
        is_radical: Tag = Tag.from_name(f"{f_kanji}is-radical")
        is_radical_purely: Tag = Tag.from_name(f"{f_kanji}is-radical-purely")
        is_radical_silent: Tag = Tag.from_name(f"{f_kanji}is-radical-silent")
        in_vocab_main_form: Tag = Tag.from_name(f"{f_kanji}in-vocab-main-form")
        in_any_vocab_form: Tag = Tag.from_name(f"{f_kanji}in-any-vocab-form")

        with_single_kanji_vocab: Tag = Tag.from_name(f"{f_kanji}single-kanji-vocab")
        with_single_kanji_vocab_with_different_reading: Tag = Tag.from_name(f"{f_kanji}single-kanji-vocab-with-different-reading")
        with_studying_single_kanji_vocab_with_different_reading: Tag = Tag.from_name(f"{f_kanji}studying-single-kanji-vocab-with-different-reading")
        with_no_primary_on_readings: Tag = Tag.from_name(f"{f_kanji}no-primary-on-readings")
        with_no_primary_readings: Tag = Tag.from_name(f"{f_kanji}no-primary-readings")
        with_studying_vocab: Tag = Tag.from_name(f"{f_kanji}studying-vocab")
        with_vocab_with_primary_on_reading: Tag = Tag.from_name(f"{f_kanji}has-vocab-with-primary-on-reading")
        with_studying_vocab_with_primary_on_reading: Tag = Tag.from_name(f"{f_kanji}studying-vocab-with-primary-on-reading")

        has_studying_vocab_with_no_matching_primary_reading: Tag = Tag.from_name(f"{f_kanji}has-studying-vocab-with-no-matching-primary-reading")
        has_studying_vocab_for_each_primary_reading: Tag = Tag.from_name(f"{f_kanji}has-studying-vocab-for-each-primary-reading")
        has_primary_reading_with_no_studying_vocab: Tag = Tag.from_name(f"{f_kanji}has-primary-reading-with-no-studying-vocab")
        has_non_primary_on_reading_vocab: Tag = Tag.from_name(f"{f_kanji}has-non-primary-on-reading-vocab")
        has_non_primary_on_reading_vocab_with_only_known_kanji: Tag = Tag.from_name(f"{f_kanji}has-non-primary-on-reading-vocab-with-only-known-kanji")

    class Vocab(Slots):
        root: str = f_vocab  # Keep as string for startswith() checks
        has_no_studying_sentences: Tag = Tag.from_name(f"{f_vocab}has-no-studying-sentences")
        question_overrides_form: Tag = Tag.from_name(f"{f_vocab}question-overrides-form")
        is_compositionally_transparent_compound: Tag = Tag.from_name(f"{f_vocab}is-compositionally-transparent-compound")
        is_ichidan_hiding_godan_potential: Tag = Tag.from_name(f"{f_vocab}is-ichidan-hiding-godan-potential")

        class Register(Slots):
            polite: Tag = Tag.from_name(f"{f_vocab_register}polite")
            formal: Tag = Tag.from_name(f"{f_vocab_register}formal")
            humble: Tag = Tag.from_name(f"{f_vocab_register}humble")
            honorific: Tag = Tag.from_name(f"{f_vocab_register}honorific")
            informal: Tag = Tag.from_name(f"{f_vocab_register}informal")
            slang: Tag = Tag.from_name(f"{f_vocab_register}slang")
            rough_masculine: Tag = Tag.from_name(f"{f_vocab_register}rough")
            soft_feminine: Tag = Tag.from_name(f"{f_vocab_register}soft")
            derogatory: Tag = Tag.from_name(f"{f_vocab_register}derogatory")
            vulgar: Tag = Tag.from_name(f"{f_vocab_register}vulgar")
            archaic: Tag = Tag.from_name(f"{f_vocab_register}archaic")
            sensitive: Tag = Tag.from_name(f"{f_vocab_register}sensitive")
            childish: Tag = Tag.from_name(f"{f_vocab_register}childish")
            literary: Tag = Tag.from_name(f"{f_vocab_register}literary")

        class Matching(Slots):
            yield_last_token_to_overlapping_compound: Tag = Tag.from_name(f"{f_vocab_matching}yield-last-token-to-upcoming-compound")
            is_poison_word: Tag = Tag.from_name(f"{f_vocab_matching}is-poison-word")
            is_inflecting_word: Tag = Tag.from_name(f"{f_vocab_matching}is-inflecting-word")

            class Requires(Slots):
                folder_name: str = f_vocab_matching_requires
                masu_stem: Tag = Tag.from_name(f"{f_vocab_matching_requires}masu_stem")
                godan: Tag = Tag.from_name(f"{f_vocab_matching_requires}godan")
                ichidan: Tag = Tag.from_name(f"{f_vocab_matching_requires}ichidan")
                irrealis: Tag = Tag.from_name(f"{f_vocab_matching_requires}irrealis")
                preceding_adverb: Tag = Tag.from_name(f"{f_vocab_matching_requires}preceding_adverb")
                past_tense_stem: Tag = Tag.from_name(f"{f_vocab_matching_requires}past-tense-stem")
                dictionary_form_stem: Tag = Tag.from_name(f"{f_vocab_matching_requires}dictionary_form_stem")
                dictionary_form_prefix: Tag = Tag.from_name(f"{f_vocab_matching_requires}dictionary_form_prefix")
                ichidan_imperative: Tag = Tag.from_name(f"{f_vocab_matching_requires}ichidan_imperative")
                godan_potential: Tag = Tag.from_name(f"{f_vocab_matching_requires}godan_potential")
                godan_imperative: Tag = Tag.from_name(f"{f_vocab_matching_requires}godan_imperative")
                godan_imperative_prefix: Tag = Tag.from_name(f"{f_vocab_matching_forbids}godan_imperative_prefix")
                te_form_stem: Tag = Tag.from_name(f"{f_vocab_matching_requires}te-form-stem")
                te_form_prefix: Tag = Tag.from_name(f"{f_vocab_matching_requires}te-form-prefix")
                sentence_end: Tag = Tag.from_name(f"{f_vocab_matching_requires}sentence-end")
                sentence_start: Tag = Tag.from_name(f"{f_vocab_matching_requires}sentence-start")
                surface: Tag = Tag.from_name(f"{f_vocab_matching_requires}surface")
                single_token: Tag = Tag.from_name(f"{f_vocab_matching_requires}single-token")
                compound: Tag = Tag.from_name(f"{f_vocab_matching_requires}compound")

            class Forbids(Slots):
                masu_stem: Tag = Tag.from_name(f"{f_vocab_matching_forbids}masu_stem")
                godan: Tag = Tag.from_name(f"{f_vocab_matching_forbids}godan")
                ichidan: Tag = Tag.from_name(f"{f_vocab_matching_forbids}ichidan")
                irrealis: Tag = Tag.from_name(f"{f_vocab_matching_forbids}irrealis")
                preceding_adverb: Tag = Tag.from_name(f"{f_vocab_matching_forbids}preceding_adverb")
                past_tense_stem: Tag = Tag.from_name(f"{f_vocab_matching_forbids}past-tense-stem")
                dictionary_form_stem: Tag = Tag.from_name(f"{f_vocab_matching_forbids}dictionary_form_stem")
                dictionary_form_prefix: Tag = Tag.from_name(f"{f_vocab_matching_forbids}dictionary_form_prefix")
                ichidan_imperative: Tag = Tag.from_name(f"{f_vocab_matching_forbids}ichidan_imperative")
                godan_potential: Tag = Tag.from_name(f"{f_vocab_matching_forbids}godan_potential")
                godan_imperative: Tag = Tag.from_name(f"{f_vocab_matching_forbids}godan_imperative")
                godan_imperative_prefix: Tag = Tag.from_name(f"{f_vocab_matching_forbids}godan_imperative_prefix")
                te_form_stem: Tag = Tag.from_name(f"{f_vocab_matching_forbids}te-form-stem")
                te_form_prefix: Tag = Tag.from_name(f"{f_vocab_matching_forbids}te-form-prefix")
                sentence_end: Tag = Tag.from_name(f"{f_vocab_matching_forbids}sentence-end")
                sentence_start: Tag = Tag.from_name(f"{f_vocab_matching_forbids}sentence-start")
                surface: Tag = Tag.from_name(f"{f_vocab_matching_forbids}surface")
                auto_yielding: Tag = Tag.from_name(f"{f_vocab_matching_forbids}auto_yielding")

            class Todo(Slots):
                with_preceding_vowel: Tag = Tag.from_name(f"{f_vocab_matching_todo}match-with-preceding-vowel")

            class Uses(Slots):
                prefix_is_not: Tag = Tag.from_name(f"{f_vocab_matching_uses}prefix-is-not")
                suffix_is_not: Tag = Tag.from_name(f"{f_vocab_matching_uses}suffix-is-not")
                required_prefix: Tag = Tag.from_name(f"{f_vocab_matching_uses}required-prefix")
                surface_is_not: Tag = Tag.from_name(f"{f_vocab_matching_uses}surface-is-not")

    priority_folder: str = f"{f_root}priority::"  # Keep as string for startswith() checks

    class Source(Slots):
        folder: str = f_source  # Keep as string for startswith() checks
        immersion_kit: Tag = Tag.from_name(f"{f_source}immersion_kit")
        jamdict: Tag = Tag.from_name(f"{f_source}jamdict")

    DisableKanaOnly: Tag = Tag.from_name("_disable_uk")
    UsuallyKanaOnly: Tag = Tag.from_name("_uk")
    TTSAudio: Tag = Tag.from_name("_tts_audio")

File: src/src_dotnet/JAStudio.Core/Note/Tags.cs
// ReSharper disable MemberCanBeInternal todo: review
namespace JAStudio.Core.Note;

public static class Tags
{
   const string FRoot = "-::";
   const string FSentence = FRoot + "sentence::";
   const string FKanji = FRoot + "kanji::";
   const string FSentenceUses = FSentence + "uses::";
   const string FVocab = FRoot + "vocab::";
   const string FVocabMatching = FVocab + "matching::";
   const string FVocabRegister = FVocab + "register::";
   const string FVocabMatchingRequires = FVocabMatching + "requires::";
   const string FVocabMatchingForbids = FVocabMatching + "forbids::";
   const string FVocabMatchingTodo = FVocabMatching + "todo::";
   const string FVocabMatchingUses = FVocabMatching + "uses::";
   const string FSource = "source::";

   public static class Sentence
   {
      public static class Uses
      {
         public static readonly Tag IncorrectMatches = Tag.FromName(FSentenceUses + "incorrect-matches");
         public static readonly Tag HiddenMatches = Tag.FromName(FSentenceUses + "hidden-matches");
      }
   }

   public static class Kanji
   {
      public static readonly Tag IsRadical = Tag.FromName(FKanji + "is-radical");
      public static readonly Tag IsRadicalPurely = Tag.FromName(FKanji + "is-radical-purely");
      public static readonly Tag IsRadicalSilent = Tag.FromName(FKanji + "is-radical-silent");
      public static readonly Tag InVocabMainForm = Tag.FromName(FKanji + "in-vocab-main-form");
      public static readonly Tag InAnyVocabForm = Tag.FromName(FKanji + "in-any-vocab-form");

      public static readonly Tag WithSingleKanjiVocab = Tag.FromName(FKanji + "single-kanji-vocab");
      public static readonly Tag WithSingleKanjiVocabWithDifferentReading = Tag.FromName(FKanji + "single-kanji-vocab-with-different-reading");
      public static readonly Tag WithStudyingSingleKanjiVocabWithDifferentReading = Tag.FromName(FKanji + "studying-single-kanji-vocab-with-different-reading");
      public static readonly Tag WithNoPrimaryOnReadings = Tag.FromName(FKanji + "no-primary-on-readings");
      public static readonly Tag WithNoPrimaryReadings = Tag.FromName(FKanji + "no-primary-readings");
      public static readonly Tag WithStudyingVocab = Tag.FromName(FKanji + "studying-vocab");
      public static readonly Tag WithVocabWithPrimaryOnReading = Tag.FromName(FKanji + "has-vocab-with-primary-on-reading");
      public static readonly Tag WithStudyingVocabWithPrimaryOnReading = Tag.FromName(FKanji + "studying-vocab-with-primary-on-reading");

      public static readonly Tag HasStudyingVocabWithNoMatchingPrimaryReading = Tag.FromName(FKanji + "has-studying-vocab-with-no-matching-primary-reading");
      public static readonly Tag HasStudyingVocabForEachPrimaryReading = Tag.FromName(FKanji + "has-studying-vocab-for-each-primary-reading");
      public static readonly Tag HasPrimaryReadingWithNoStudyingVocab = Tag.FromName(FKanji + "has-primary-reading-with-no-studying-vocab");
      public static readonly Tag HasNonPrimaryOnReadingVocab = Tag.FromName(FKanji + "has-non-primary-on-reading-vocab");
      public static readonly Tag HasNonPrimaryOnReadingVocabWithOnlyKnownKanji = Tag.FromName(FKanji + "has-non-primary-on-reading-vocab-with-only-known-kanji");
   }

   public static class Vocab
   {
      public const string Root = FVocab; // Keep as string for startswith() checks

      public static readonly Tag HasNoStudyingSentences = Tag.FromName(FVocab + "has-no-studying-sentences");
      public static readonly Tag QuestionOverridesForm = Tag.FromName(FVocab + "question-overrides-form");
      public static readonly Tag IsCompositionallyTransparentCompound = Tag.FromName(FVocab + "is-compositionally-transparent-compound");
      public static readonly Tag IsIchidanHidingGodanPotential = Tag.FromName(FVocab + "is-ichidan-hiding-godan-potential");

      public static class Register
      {
         public static readonly Tag Polite = Tag.FromName(FVocabRegister + "polite");
         public static readonly Tag Formal = Tag.FromName(FVocabRegister + "formal");
         public static readonly Tag Humble = Tag.FromName(FVocabRegister + "humble");
         public static readonly Tag Honorific = Tag.FromName(FVocabRegister + "honorific");
         public static readonly Tag Informal = Tag.FromName(FVocabRegister + "informal");
         public static readonly Tag Slang = Tag.FromName(FVocabRegister + "slang");
         public static readonly Tag RoughMasculine = Tag.FromName(FVocabRegister + "rough");
         public static readonly Tag SoftFeminine = Tag.FromName(FVocabRegister + "soft");
         public static readonly Tag Derogatory = Tag.FromName(FVocabRegister + "derogatory");
         public static readonly Tag Vulgar = Tag.FromName(FVocabRegister + "vulgar");
         public static readonly Tag Archaic = Tag.FromName(FVocabRegister + "archaic");
         public static readonly Tag Sensitive = Tag.FromName(FVocabRegister + "sensitive");
         public static readonly Tag Childish = Tag.FromName(FVocabRegister + "childish");
         public static readonly Tag Literary = Tag.FromName(FVocabRegister + "literary");
      }

      public static class Matching
      {
         public static readonly Tag YieldLastTokenToOverlappingCompound = Tag.FromName(FVocabMatching + "yield-last-token-to-upcoming-compound");
         public static readonly Tag IsPoisonWord = Tag.FromName(FVocabMatching + "is-poison-word");
         public static readonly Tag IsInflectingWord = Tag.FromName(FVocabMatching + "is-inflecting-word");

         public static class Requires
         {
            public const string FolderName = FVocabMatchingRequires;

            public static readonly Tag MasuStem = Tag.FromName(FVocabMatchingRequires + "masu_stem");
            public static readonly Tag Godan = Tag.FromName(FVocabMatchingRequires + "godan");
            public static readonly Tag Ichidan = Tag.FromName(FVocabMatchingRequires + "ichidan");
            public static readonly Tag Irrealis = Tag.FromName(FVocabMatchingRequires + "irrealis");
            public static readonly Tag PrecedingAdverb = Tag.FromName(FVocabMatchingRequires + "preceding_adverb");
            public static readonly Tag PastTenseStem = Tag.FromName(FVocabMatchingRequires + "past-tense-stem");
            public static readonly Tag DictionaryFormStem = Tag.FromName(FVocabMatchingRequires + "dictionary_form_stem");
            public static readonly Tag DictionaryFormPrefix = Tag.FromName(FVocabMatchingRequires + "dictionary_form_prefix");
            public static readonly Tag IchidanImperative = Tag.FromName(FVocabMatchingRequires + "ichidan_imperative");
            public static readonly Tag GodanPotential = Tag.FromName(FVocabMatchingRequires + "godan_potential");
            public static readonly Tag GodanImperative = Tag.FromName(FVocabMatchingRequires + "godan_imperative");
            public static readonly Tag GodanImperativePrefix = Tag.FromName(FVocabMatchingForbids + "godan_imperative_prefix");
            public static readonly Tag TeFormStem = Tag.FromName(FVocabMatchingRequires + "te-form-stem");
            public static readonly Tag TeFormPrefix = Tag.FromName(FVocabMatchingRequires + "te-form-prefix");
            public static readonly Tag SentenceEnd = Tag.FromName(FVocabMatchingRequires + "sentence-end");
            public static readonly Tag SentenceStart = Tag.FromName(FVocabMatchingRequires + "sentence-start");
            public static readonly Tag Surface = Tag.FromName(FVocabMatchingRequires + "surface");
            public static readonly Tag SingleToken = Tag.FromName(FVocabMatchingRequires + "single-token");
            public static readonly Tag Compound = Tag.FromName(FVocabMatchingRequires + "compound");
         }

         public static class Forbids
         {
            public static readonly Tag MasuStem = Tag.FromName(FVocabMatchingForbids + "masu_stem");
            public static readonly Tag Godan = Tag.FromName(FVocabMatchingForbids + "godan");
            public static readonly Tag Ichidan = Tag.FromName(FVocabMatchingForbids + "ichidan");
            public static readonly Tag Irrealis = Tag.FromName(FVocabMatchingForbids + "irrealis");
            public static readonly Tag PrecedingAdverb = Tag.FromName(FVocabMatchingForbids + "preceding_adverb");
            public static readonly Tag PastTenseStem = Tag.FromName(FVocabMatchingForbids + "past-tense-stem");
            public static readonly Tag DictionaryFormStem = Tag.FromName(FVocabMatchingForbids + "dictionary_form_stem");
            public static readonly Tag DictionaryFormPrefix = Tag.FromName(FVocabMatchingForbids + "dictionary_form_prefix");
            public static readonly Tag IchidanImperative = Tag.FromName(FVocabMatchingForbids + "ichidan_imperative");
            public static readonly Tag GodanPotential = Tag.FromName(FVocabMatchingForbids + "godan_potential");
            public static readonly Tag GodanImperative = Tag.FromName(FVocabMatchingForbids + "godan_imperative");
            public static readonly Tag GodanImperativePrefix = Tag.FromName(FVocabMatchingForbids + "godan_imperative_prefix");
            public static readonly Tag TeFormStem = Tag.FromName(FVocabMatchingForbids + "te-form-stem");
            public static readonly Tag TeFormPrefix = Tag.FromName(FVocabMatchingForbids + "te-form-prefix");
            public static readonly Tag SentenceEnd = Tag.FromName(FVocabMatchingForbids + "sentence-end");
            public static readonly Tag SentenceStart = Tag.FromName(FVocabMatchingForbids + "sentence-start");
            public static readonly Tag Surface = Tag.FromName(FVocabMatchingForbids + "surface");
            public static readonly Tag AutoYielding = Tag.FromName(FVocabMatchingForbids + "auto_yielding");
         }

         public static class Todo
         {
            public static readonly Tag WithPrecedingVowel = Tag.FromName(FVocabMatchingTodo + "match-with-preceding-vowel");
         }

         public static class Uses
         {
            public static readonly Tag PrefixIsNot = Tag.FromName(FVocabMatchingUses + "prefix-is-not");
            public static readonly Tag SuffixIsNot = Tag.FromName(FVocabMatchingUses + "suffix-is-not");
            public static readonly Tag RequiredPrefix = Tag.FromName(FVocabMatchingUses + "required-prefix");
            public static readonly Tag SurfaceIsNot = Tag.FromName(FVocabMatchingUses + "surface-is-not");
         }
      }
   }

   public const string PriorityFolder = FRoot + "priority::"; // Keep as string for startswith() checks

   public static class Source
   {
      public const string Folder = FSource; // Keep as string for startswith() checks

      public static readonly Tag ImmersionKit = Tag.FromName(FSource + "immersion_kit");
      public static readonly Tag Jamdict = Tag.FromName(FSource + "jamdict");
   }

   public static readonly Tag DisableKanaOnly = Tag.FromName("_disable_uk");
   public static readonly Tag UsuallyKanaOnly = Tag.FromName("_uk");
   public static readonly Tag TTSAudio = Tag.FromName("_tts_audio");
}
